feat: add service support to analytics mechnism (#1938)
AmadeusITGroup/otter: docs/analytics/PERFORMANCE.md

This has to be called only once in a single page application, as it is only meaningful for the initial page load. [FirstLoadDataPayload](https://github.com/AmadeusITGroup/otter/blob/main/packages/@o3r/analytics/src/contracts/events-contracts.ts)
interface is the model object for this mark.

### First paint ([FP](https://developers.google.com/web/tools/lighthouse/audits/first-contentful-paint))

This is one of the first metrics for perceived performance. Basically, it measures the time the app takes to answer a
user's first question: Is something happening? Is the navigation successful ? Has the server responded?
The First Paint (FP) measures the time it takes from the start of the navigation to, for example, display the loading indication.

### First Meaningful Paint ([FMP](https://developers.google.com/web/tools/lighthouse/audits/first-meaningful-paint))

Also for perceived performance, FMP measures the time the app takes to render enough content for users to engage. A simple strategy for this metric is to mark what's called hero elements (most important

> elements in the page) and register the time it took to display them

Reviewer: ```suggestion
elements in the page) and register the time it took to display them.
```

### Time to Interactive ([TTI](https://developers.google.com/web/tools/lighthouse/audits/time-to-interactive))

TTI marks the time when the user can effectively interact with the app. This is closely related to the fact that, in some implementations, the app may have rendered meaningful information
(measured by FMP) but, in the background, it's still doing some kind of computation that blocks any possible interaction with the page.

The time to interactive is quite tricky as it not only depends on the relevant data readiness, but also on
component internal display mechanics.

> If you know exactly where javascript will trigger a layout change  (e.g. by passing a boolean variable to true), it's possible to measure the upper bound for the rendering.

Reviewer: It is a bit vague, maybe an implemented example could help the user understand better here

In addition, during a component development, you can't possibly know beforehand if the component will be relevant for a TTI or not, since it depends on the page itself.

> For example, the display of a cart component may be relevant for TTI in a given page and not relevant at all in others.

Reviewer: (product selection vs product recap?)

Hence, you cannot really define your TTI logic at component level.

Given the above facts, we advise to split the TTI metric in two:

* __dataReady__: This probe marks the time when all the data, needed to the page be interactive, is available
* __TTI per component__: data ready for each component; we advise to implement it later, since it may impact the complexity of the code

> For the time being we will consider only the implementation of __data ready__

Reviewer: ```suggestion
For the time being we will consider only the implementation of __data ready__.
```

### Network and server-side metrics

As the browser can't understand when a route event happens in an SPA, the NavigationTimingAPI can't be directly used apart from the first page load at most.

> Subsequent routing changes won't profit of the API connection timings.

Reviewer: ```suggestion
Subsequent routing changes won't profit from the API connection timings.
```

> In regard of the __server fetches__ (filter out from the resource timing API), the [PerformanceMetricPlugin](https://github.com/AmadeusITGroup/otter/blob/main/packages/@ama-sdk/core/src/plugins/perf-metric/perf-metric.fetch.ts)

Reviewer: ```suggestion
In regard of the __server fetches__ (filtered out from the resource timing API), the [PerformanceMetricPlugin](https://github.com/AmadeusITGroup/otter/blob/main/packages/@ama-sdk/core/src/plugins/perf-metric/perf-metric.fetch.ts)
```
@sdo-1A : could you confirm this one please?

has been put in place to get the metrics associated to server calls.
Check [ServerCallMetric](https://github.com/AmadeusITGroup/otter/blob/main/packages/@o3r/analytics/src/contracts/events-contracts.ts)
model to see which information is saved for each call.

## How to mark performance metrics?

> The __EventTrackService__ plugs itself to the [NavigationEnd](https://angular.io/api/router/NavigationEnd) router, to handle the performance metrics and exposes the performance object as a stream (observable).

Reviewer: ```suggestion
The __EventTrackService__ plugs itself to the [NavigationEnd](https://angular.io/api/router/NavigationEnd) router, to handle the performance metrics and expose the performance object as a stream (observable).
```
or 
```suggestion
The __EventTrackService__ plugs itself to the [NavigationEnd](https://angular.io/api/router/NavigationEnd) router to handle the performance metrics, and exposes the performance object as a stream (observable).
```

The performance metric object structure is defined by __PerfEventPayload__ interface which can be found [here](https://github.com/AmadeusITGroup/otter/blob/main/packages/@o3r/analytics/src/contracts/events-contracts.ts).
The service provides a way to activate/deactivate performance measurements. By default, it's __activated__ and we expose a public method called __togglePerfTracking__ to activate/deactivate it.
For instance if you want to deactivate it, call this in your app:

```typescript
import {EventTrackService} from '@o3r/analytics';
...
constructor(trackService: EventTrackService) {
  trackService.togglePerfTracking(false);
}
```

### Tracking configuration

You can override the default configuration via a configuration token ([EVENT_TRACK_SERVICE_CONFIGURATION](https://github.com/AmadeusITGroup/otter/blob/main/packages/@o3r/analytics/src/services/event-track/event-track.configuration.ts)).
Example of configuration override:

```typescript
// in app module
  ...
  providers: [
    ...
    {provide: EVENT_TRACK_SERVICE_CONFIGURATION, useValue: {useBrowserApiForFirstFP: true}}
  ]
```

> More details about the configuration object and [defaultEventTrackConfiguration](https://github.com/AmadeusITGroup/otter/blob/main/packages/@o3r/analytics/src/services/event-track/event-track.configuration.ts) can be found [here](https://github.com/AmadeusITGroup/otter/blob/main/packages/@o3r/analytics/src/services/event-track/event-track.configuration.ts)

Reviewer: ```suggestion
More details about the configuration object and [defaultEventTrackConfiguration](https://github.com/AmadeusITGroup/otter/blob/main/packages/@o3r/analytics/src/services/event-track/event-track.configuration.ts) can be found [in the event tracking documentation](https://github.com/AmadeusITGroup/otter/blob/main/packages/@o3r/analytics/src/services/event-track/event-track.configuration.ts)
```
NB: `here` is too vague to be the handle for a link  (for screen reader or any tool that would read metadata from the documentation).
